### PR TITLE
network, fqdn: Fix adding subdomain search entry flow

### PR DESCRIPTION
**Pull request by oshoval**
**What this PR does / why we need it**:

The default virt-launcher `dnsPolicy` is `ClusterFirst`
(unless changed via the VMI spec).
It means that k8s default search domains will be added as the first ones,
and in case the nodes have more entries, they will be appended
to the pod's search list. [1]

The current logic of adding the missing subdomain entry,
is to find the longest entry, and to add the subdomain to this entry.

In case the longest entry will be a node entry instead of the
entries that k8s added itself, the logic would fail to find
and add the subdomain to the right service domain entry.

The fix is to find the longest svc entry instead of the absolute longest entry,
which indicates it is a k8s added service entry, and to prepend this entry
with the subdomain if needed.

We search for `.svc.` because it means it is a service added by k8s itself.
Only the longest service entry is the one that we should add subdomain to,
so it will be of the form `subdomain.namespace.svc.cluster.local`
(where cluster.local is the cluster domain).
See [2] for additional info.

In case no service entry was found a subdomain won't be added.
One example is if `dnsPolicy` is `None` and no service domain included manually.

Note:
The fix affects DHCP option 119 (domain search list).
The expected unchanged behavior is that option 15 still uses the longest entry
from the domain search list.

[1] https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-policy
[2] https://kubernetes.io/docs/tasks/debug-application-cluster/debug-service/

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
None
```

### CODE REVIEW

**Review by EdDev — COMMENTED** [review]
I do not understand why 4 e2e tests have been added here.
I would expect to see this on existing tests and not raise the cost bar so high.

File: tests/network/vmi_subdomain.go [hunk 1]
```diff
@@ -50,9 +51,13 @@ var _ = SIGDescribe("Subdomain", func() {
 		tests.BeforeTestCleanup()
 	})
 
+	const (
+		subdomain     = "testsubdomain"
+		k8sNameserver = "10.96.0.10"
```
[the remark this answers is not in the document]
Comment (oshoval) [reply tests/network/vmi_subdomain.go:56]: It is the default Ipv4 k8s cluster dns name server

File: pkg/network/dns/resolveconf.go [hunk 2]
```diff
@@ -84,6 +84,17 @@ func GetDomainName(searchDomains []string) string {
 	return selected
 }
 
+//GetLongestServiceDomainName returns the longest service search domain entry
+func GetLongestServiceDomainName(searchDomains []string) string {
```
[the remark this answers is not in the document]
Comment (oshoval) [reply pkg/network/dns/resolveconf.go:88]: yes, for example option 15 still takes the absolute longest out of the search domain list

File: pkg/network/dns/resolveconf.go [hunk 3]
```diff
@@ -96,8 +107,8 @@ func DomainNameWithSubdomain(searchDomains []string, subdomain string) string {
 		return ""
 	}
 
-	domainName := GetDomainName(searchDomains)
-	if !strings.Contains(domainName, subdomain) {
+	domainName := GetLongestServiceDomainName(searchDomains)
+	if domainName != "" && !strings.HasPrefix(domainName, subdomain+".") {
```
[the remark this answers is not in the document]
Comment (oshoval) [reply pkg/network/dns/resolveconf.go:111]: yep
https://github.com/kubevirt/kubevirt/pull/6985/files#diff-2fd511a3fa9950457ec29c983210a31f813a1634643b91c8435b1b964b22597fR152

**Comment by oshoval** [comment]
> I do not understand why 4 e2e tests have been added here. I would expect to see this on existing tests and not raise the cost bar so high.

the previous ones checked fqdn with the default pod config,
the added ones are 
bridge / masqurade when there is a custon dnsPolicy, with a long interface
bridge / masqurade when there is a custon dnsPolicy, with no svc at all

we need those in order to create a specific dns policy and check the new updated flow
but i can remove masqurade and leave only bridge on those two, because the code is the same for both (dhcp and the changed  funtion are common for both masqurade and bridge)

File: pkg/network/dns/resolveconf.go [hunk 4]
```diff
@@ -84,6 +84,17 @@ func GetDomainName(searchDomains []string) string {
 	return selected
 }
 
+//GetLongestServiceDomainName returns the longest service search domain entry
+func GetLongestServiceDomainName(searchDomains []string) string {
+	selected := ""
+	for _, d := range searchDomains {
+		if len(d) > len(selected) && strings.Contains(d, ".svc.") {
```
[the remark this answers is not in the document]
Comment (oshoval) [reply pkg/network/dns/resolveconf.go:91]: > Please place the `svc` key into a const and ref from where it came.

will add
refereed in PR desc if thats what you mean

> BTW, this is a bit unsafe. Someone may add this as a subdomain by itself or it can be part of the domain regardless.

Right (not as part of the subdomain because the subdomain is the first, and the code here has dot before `svc` as wel, but as part of the domain theoretically)
I think its not a good practice to have it as part of the domain,
we can make it bit more stricter if we will need
Note that even getting the cluster domain doesn't have a clean API method
https://github.com/kubernetes/website/issues/25758  (closed without solution)

File: pkg/network/dns/resolveconf_test.go [hunk 5]
```diff
@@ -92,8 +92,23 @@ var _ = Describe("Resolveconf", func() {
...
+				"t4lanpt7z4ix58nvxl4d.com",
+				"14wg5xngig6vzfqjww4kocnky3c9dqjpwkewzlwpf.com",
+				"4wg5xngig6vzfqjww4kocnky3c9dqjpwkewzlwpf.com",
+				"this.is.a.svc.service.domain",
+			}
+			domain := GetLongestServiceDomainName(searchDomains)
+			Expect(domain).To(Equal("this.is.a.svc.service.domain"))
+		})
+	})
+
```
[the remark this answers is not in the document]
Comment (oshoval) [reply pkg/network/dns/resolveconf_test.go:109]: https://github.com/kubevirt/kubevirt/pull/6985/files/a5dc1208fb28adb99db606edf1f9911f25635c52#diff-2fd511a3fa9950457ec29c983210a31f813a1634643b91c8435b1b964b22597fR143
this test already checks it via `DomainNameWithSubdomain` which is the `GetLongestServiceDomainName` caller
I will remove the test you mention (just added it so it will be a concrete unit test of the new function)

File: tests/network/vmi_subdomain.go [hunk 6]
```diff
@@ -106,3 +154,12 @@ func fedoraBridgeBindingVMI() *v1.VirtualMachineInstance {
 		libvmi.WithInterface(libvmi.InterfaceDeviceWithBridgeBinding(libvmi.DefaultInterfaceName)),
 		libvmi.WithNetwork(v1.DefaultPodNetwork()))
 }
+
+func verifyFQDN(vmi *v1.VirtualMachineInstance, expectedFQDN string) error {
```
[the remark this answers is not in the document]
Comment (oshoval) [reply tests/network/vmi_subdomain.go:158]: it doesn't assert (Expect), it only returns an `error` type and later asserted in the test itself, so i will rename it to 
`hasExpectedFQDN` ?

**Comment by oshoval** [comment]
addressed comments

File: tests/network/vmi_subdomain.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (EdDev) [reply tests/network/vmi_subdomain.go:56]: We do not always control the cluster on which these tests run against.
I do not think you can assume it.

Perhaps you can read it from some other pod or just invent one with a link-local address.

[the remark this answers is not in the document]
Comment (oshoval) [reply tests/network/vmi_subdomain.go:56]: maybe instead of this, we can change the test then to only check the expected resolv.conf instead `hostname -f` ?
(after all this what should be checked by this test, the rest might be additionals)
and then we can use this ip but it won't actually be used.
This what happens as well on other test that uses `DNSPolicy` None
and it will keep it simple

[the remark this answers is not in the document]
Comment (EdDev) [reply tests/network/vmi_subdomain.go:56]: This is an e2e test, we are interested to check all the path if possible. I am sure there is a way to read this value using client-go or alike. I think we do have code that checks services already, no?

[the remark this answers is not in the document]
Comment (oshoval) [reply tests/network/vmi_subdomain.go:56]: I just verified on Downstream cluster, they don't have the same service as we have,
so if we read it, we better read it from a pod, for example virt-handler that always should exists

**Comment by oshoval** [comment]
updated tests to use `ClusterDNSServiceIP` in order to determine cluster dns ip

File: tests/network/vmi_subdomain.go [hunk 7]
```diff
@@ -80,18 +83,62 @@ var _ = SIGDescribe("Subdomain", func() {
...
+
+			vmi, err := virtClient.VirtualMachineInstance(util.NamespaceTestDefault).Create(vmiSpec)
+			Expect(err).ToNot(HaveOccurred())
+			vmi = tests.WaitUntilVMIReady(vmi, console.LoginToFedora)
+
+			Expect(assertFQDNinGuest(vmi, expectedFQDN)).To(Succeed(), "failed to get expected FQDN")
+		})
+	})
+
+	It("VMI with custom DNSPolicy and a subdomain should have the expected resolv.conf", func() {
```
Comment (EdDev) [inline tests/network/vmi_subdomain.go:118]: It took me some time to understand that this test is having a subdomain configured on the VMI but as there is no headless service, it is not expected to be added to the guest. Please confirm that I understand correctly.
Please try to make this clearer in the text.

The second point I have, is about the differences from the other test you added.
This one uses `resolv.conf` while the other`hostname` to check for the FQDN. What is the reason for the difference?
Also, consider extracting the assertion here to a function like the other one.

**Review by EdDev — COMMENTED** [review]
Looks good, thank you!
I only have one inline comment regarding the e2e tests and a nit about the commit message.

Commit message:
Can you check that you are not passing 72 chars? It seems you do and it will be nicer if it is wrapped.

> The fix affects DHCP option 119 (domain search list) and option 15 (domain name).
> The expected unchanged behavior is the option 15 still uses the longest entry from the domain search list.

Regarding option 15 I am a bit confused. On one hand you mention the fix affects it and on the other hand that the behavior is unchanged. Can you please clarify?
Also, the last sentence above needs some fixing, try to rework it.

**Comment by oshoval** [comment]
> Regarding option 15 I am a bit confused. On one hand you mention the fix affects it and on the other hand that the behavior is unchanged. Can you please clarify?

The PR only fix that if subdomain will be added, it will add only to the right entry.
I added this note so it the flow will be clearer for the future reader.

1. We read pod's `/etc/resolv.conf` to a list.
2. Check if we need to add subdomain entry (this PR fix this logic), if so add to the list.
3. The list is propagated as DHCP option 119.
4. Option 15 takes the longest from option 119.

One thing that point 2 fixed is that if you have a subdomain, but don't have any service entry
(used `dnsPolicy` `None` and entered manually the domains),
the previous logic would add subdomain to an entry in a wrong manner, and it will also reflected on option 15
because option 15 takes the longest that might changed.
For example if `/etc/resolv.conf` was `example.com` it would changed to `subdomain.example.com`
This why I added the e2e test for it.

The bottom line is that the logic of option 15 was not changed, it still takes the longest.
A follow discussion might be what option 15 logic should be, but not as part of this PR.

File: tests/network/vmi_subdomain.go (hunk 7 above)
Comment (oshoval) [reply tests/network/vmi_subdomain.go:118]: > It took me some time to understand that this test is having a subdomain configured on the VMI but as there is no headless service, it is not expected to be added to the guest. Please confirm that I understand correctly.

Exactly, as we can't update any entry with subdomain, just specific ones.

> The second point I have, is about the differences from the other test you added. This one uses `resolv.conf` while the other`hostname` to check for the FQDN. What is the reason for the difference? 

You can't check FQDN when you don't have the needed entries (`subdomain.namespace.svc.cluster.local`), it will be just empty or equal VM name.
So this tests checks the exact thing it should which is that the guest `/etc/resolv.conf` has the expected `search` line

Comment (EdDev) [reply tests/network/vmi_subdomain.go:118]: Thank you for the clarification.

I think it will be beneficial if the test name/text expresses what is expected here given the state that affects it most.
Requiring one to delve into the test code to understand basic expectations is not nice.

What do you think about this: `VMI with custom DNSPolicy, a subdomain and no service entry, should not include the subdomain in the searchlist`.

Regarding testing the `hostname` and `resolv.conf`: Sorry, I still do not understand why you could not use the same. On one you do get what you expect, on the other you do not (expect not to get it). You could also perform both checks if you think it is better. My main point is that having two ways to check something is confusing and not consistent, it will be better to show we are checking the same with different expectations.

**Comment by EdDev** [comment]
> The bottom line is that the logic of option 15 was not changed, it still takes the longest. A follow discussion might be what option 15 logic should be, but not as part of this PR.

Ok, so just remove the note about fixing also option 15, as that part did not changed.

File: tests/network/vmi_subdomain.go (hunk 7 above)
Comment (oshoval) [reply tests/network/vmi_subdomain.go:118]: Adding the expect not to get it, 
Note that there is also no headless service there, so it is not the only reason we won't get it

**Comment by oshoval** [comment]
> Ok, so just remove the note about fixing also option 15, as that part did not changed.

Done, 
and pushed rest of the changes + the minor commit message changes and such.

**Review by EdDev — APPROVED** [review]
Thank you!

**Comment by oshoval** [comment]
Thanks Edy
@AlonaKaplan 
can you please take a look ?
